modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/editor/BaseEditor.js

```diff
@@ -0,0 +1,100 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import PropTypes from 'prop-types';
+import React, {forwardRef, useCallback, useEffect, useRef} from 'react';
+
+import {Editor} from './Editor';
```

Comment:
Keeping a dependency to `Editor` is an interesting architectural proposal, but I don't think is the best way to go:
- In my view, there is no difference between `getHTML`-related code and `beforeScreenFlip` code. Both are internal DXP-related tweaks / quirks, that should be kept under the hood. Both are common to **all** variants of CKeditor usages. There is no need for two layers of base code.
- We want all variants of CKeditor to have all common DXP-specific code. The way would be in this PR, `Editor` would be missing `getHTML` stuff.
- We want to support `ClassicEditor` as a thin wrapper to CKeditor in DXP. We don't want an odd alternative with only some DXP fixes. As such, we want to deprecate `Editor`, and in a future iteration, get rid of it. If we keep dependency like in this PR, we cannot deprecate/remove it.

What we should do here is:
- `BaseEditor` should import `ckeditor4-react`.
- `beforeScreenFlip` should be in `BaseEditor`.
- `Editor` should say exactly as it is, so we don't break backwards compatibility.
- We should deprecate `Editor`.

Reply:
I like the idea of abstracting out common stuff in a `BaseEditor` and do not export it in the component `index.js`.

I'm a little bit in doubt about keeping `Editor.js` as part of all this new architecture. On one hand, "Editor" is a name that does not suggest what we wanted ("do not use it as it's an internal thing"), so it's probably better to deprecate it and add a comment in the file itself stating the intent. It's not a big deal if we write `BaseEditor` with no dependency on `Editor`.

But on the other, a proper deprecation requires usage analysis and determining whether existing uses can be replaced by some "other" thing. And that's where my doubts arise. What that other thing would look like? Some new variant on top of `BaseEditor`? or maybe just a plain, bare `ckeditor4-react`, with no support for DXP machinery (is it even a good idea)? The alleged purpose of this task is to offer a unified, consistent editor DXP API covering all the cases we know of. Could it be a potential outcome after the analysis is that we can't "deprecate" what `Editor` provides? (I'm hoping it's not)

Reply:
>But on the other, a proper deprecation requires usage analysis and determining whether existing uses can be replaced by some "other" thing. And that's where my doubts arise. What that other thing would look like? Some new variant on top of BaseEditor? or maybe just a plain, bare ckeditor4-react, with no support for DXP machinery (is it even a good idea)? The alleged purpose of this task is to offer a unified, consistent editor DXP API covering all the cases we know of. Could it be a potential outcome after the analysis is that we can't "deprecate" what Editor provides? (I'm hoping it's not)

Yeah, valid concerns! I'd say that it's almost certain that `ClassicEditor` will be perfectly fine. There is a tiny chance that is not, if usages do something related to DXP quirks in common code. We can analyze, test and migrate usages in a separate task. We should do that asap after this PR is merged.